colossalai/inference/batch_bucket.py

```diff
@@ -0,0 +1,449 @@
+
+        num_seqs_to_add = min(self.max_batch_size - self._current_batch_size, len(seqs))
+        block_tables = None
+        if num_seqs_to_add > 0:
+            for i, seq in enumerate(seqs[:num_seqs_to_add]):
+                self._sequences_dict[seq.request_id] = seq
+                self._sequences_indexes[seq.request_id] = self._current_batch_size + i
+            # TODO external (rename): modify Sequence.sentence_len to seq_len
+            self._sequence_lengths[
+                self._current_batch_size : self._current_batch_size + num_seqs_to_add
+            ] = torch.tensor([seq.sentence_len for seq in seqs[:num_seqs_to_add]], dtype=torch.int32)
+            # NOTE block tables to be updated by kvcache manager
+            block_tables = self._block_tables[self._current_batch_size : self._current_batch_size + num_seqs_to_add]
+            if alloc_block_tables is not None:
+                # copy block ids from provided block tables
+                self._block_tables[
+                    self._current_batch_size : self._current_batch_size + num_seqs_to_add
+                ] = alloc_block_tables
+            elif alloc_block_tables_fn:
+                alloc_block_tables_fn(
+                    block_tables,
+                    self._sequence_lengths[self._current_batch_size : self._current_batch_size + num_seqs_to_add],
+                )
+
+            self._current_batch_size += num_seqs_to_add
+            seqs[:] = seqs[num_seqs_to_add:]
+
+        return block_tables
+
+    def pop_seq_update_batch(
+        self, request_id: int, free_block_table_fn: Callable[[torch.Tensor], None] = None
+    ) -> Tuple[Sequence, Union[torch.Tensor, None]]:
+        """Pop a single sequence by id from the batch, and update the batch bucket status.
+
+        Args:
+            request_id (int): The uid of the sequence
+            free_block_table_fn (Callable): The function to free the block table of a sequence,
+                if not provided, then we have to release the block table manually after calling this method
+
+        Returns:
+            A tuple of: seq (Sequence): The target sequence
+            and block_table (torch.Tensor): block table of the target sequence indicating corresponding blocks,
+                none if the sequence is not found or free_block_table_fn is provided.
+        """
+        seq: Sequence = self._sequences_dict.get(request_id)
+        block_table = None
+        if seq is not None:
+            assert request_id in self._sequences_indexes, "Inconsistency in BatchBucket indexing"
+            self._sequences_dict.pop(request_id)
+            seq_b_idx = self._sequences_indexes.get(request_id)
+
+            if self.current_batch_size > 1:
+                # replace seq length of the target seq with that of the last seq in the batch
+                last_seq_b_idx = self.current_batch_size - 1
+                last_seq_id = next(
```

Comment:
can this be easier? as you only need to pop the request and free the right `sequence lengths` and `indexes`

Reply:
If only piping and freeing sequence lengths and indexes, then we will get a non-contiguous block tables which will cause errors in kernels. Image the condition that seq1, seq2, and seq3 occupy the first 3 rows in the aggregated block tables, and then we pop seq2 without replacing its space in block tables. Then we will get sequences "seq1, seq3" in the batch, but block_tables[0] -> for seq1, block_tables[2] -> for seq3,  block_tables[1] -> for ??? (an non-exist sequence).

Reply:
Understand, it's indeed necessary